### ISSUE TITLE
Add support for multiple sources in argocd cli diff|manifests

### ISSUE THREAD

**Issue by marcportabellaclotet-mt** [argoproj/argo-cd#13682]
# Summary

ArgoCD 2.6+ introduced multi source apps. Argocd CLI is not able to render manifests or show diffs, when multiple application have a different revision. 

# Motivation

Sample multiple source app:
```
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: argo-rollouts
  namespace: argocd
  labels:
    type: deployment
    application: argo-rollouts
    timestamp: lol
spec:
  project: default
  sources:
    - repoURL: 'https://argoproj.github.io/argo-helm'
      targetRevision: 2.22.3
      chart: argo-rollouts
      helm:
        valueFiles:
        - $values/argocd/applications/values/argo-rollouts/values-common.yaml
    - repoURL: https://github.com/my-repo/argo-iac.git
      targetRevision: HEAD
      ref: values
  destination:
    server: 'https://kubernetes.default.svc'
    namespace: argo-rollouts
```

In the above example, we have 2 different targetRevisions:
- One for public helm chart
- One for the values of this helm chart, which are stored in another github repo

When running argocd cli, the revision parameter refers to the targetRevision, and only one value can be provided:
```
 argocd app diff argocd/rollouts --grpc-web --revision=xxx
```

# Proposal

I think that 2 approaches could be possible to handle this:
- Be able to define target revision per source. (Not ideal)
- Render the whole application manifest based on the github branch content of the application. This means that if we have a new branch where this application is defined with same changes, use the target revision and values defined in this branch.

**Comment by cmontemuino** [comment]
I also notice the same issue without using helm.

For example:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: gitops
  namespace: argocd
  finalizers:
    - resources-finalizer.argocd.argoproj.io
spec:
  project: cluster-admin
  sources:
    - repoURL: git@gitlab.example.com:k8s/gitops.git
      targetRevision: main
      path: infra/gitops/overlays/cluster-a
    - repoURL: git@gitlab.example.com:k8s/gitops.git
      targetRevision: main
      path: infra/auth/overlays/cluster-a
  destination:
    namespace: argocd
    server: https://kubernetes.default.svc
  syncPolicy:
    automated:
      selfHeal: true
      prune: true
```

It seems like the desired state config is taken form the first source only. It does not matter if I use `argocd app diff <app-name> --revision <rev> --refresh` or `argocd app diff <app-name> --local /path-to-cloned-repo-at-specific-revision --server-side-generate`

**Comment by jtcressy** [comment]
Lack of this functionality for ArgoCD CLI is still very impactful for my organization which is currently running on v2.8.2... We have helm charts hosted in OCI repositories that use values files from an additional gitops repo and are unable to use the CLI to generate any diffs. 

This is broken even if we only need to see a diff of the chart version and leave the gitops repo revision alone. 

The API call for GetManifests on argocd-server (which in turn makes a secondary call to repo-server) always fails for us with error 500, citing that the values.yaml file is unavailable. I believe this means that #13264 is related in some way, most likely as a blocker to any enhancements for the CLI. The GetManifests route in argocd-server will have to be updated to support multiple sources and/or communicate the multiple sources to repo-server before ArgoCD CLI can support them.

Possibly, changes could be made starting with this line here:
https://github.com/argoproj/argo-cd/blob/75b438a27884bea61bda1cfe3d00d0781ba1d279/server/application/application.go#L434

argocd-server will only ever grab the first source from the application spec, as the implementation for `GetSource` shows:
https://github.com/argoproj/argo-cd/blob/75b438a27884bea61bda1cfe3d00d0781ba1d279/pkg/apis/application/v1alpha1/types.go#L208-L218

`GetSources` could be used instead, but the next step would be making sure repo-server can accept multiple sources via `ManifestRequest` in its [repository protobuf](https://github.com/argoproj/argo-cd/blob/75b438a27884bea61bda1cfe3d00d0781ba1d279/reposerver/repository/repository.proto#L10)

This is only my preliminary digging into this issue, and I haven't contributed yet otherwise i'd start working on a PR. I'll continue reading into it but if anyone can provide additional guidance it will be greatly appreciated.

**Comment by xiaoyaozibc** [comment]
Any update on this? Multi-sources is a great feature by ArgoCD, but this seems to be a bug that make the multi-sources feature less useful.

**Comment by xiaoyaozibc** [comment]
@ishitasequeira That is awesome, thank you for the PR! I noticed the same issue for the API `/api/v1/applications/{name}/manifests`. Could you please also update the API to support multi-source application? Hopefully the API is using similar code path so that it's not any extra work, or has the API been fixed already in 2.11?

**Comment by ishitasequeira** [comment]
@xiaoyaozibc , the api has been updated for `/api/v1/applications/{name}/manifests` as well. It will be available in v2.11.
